### ISSUE TITLE
Weird crashes in Json

### ISSUE THREAD

**Issue by FeepingCreature** [vibe-d/vibe.d#2205]
I got weird nonsensical crashes in my app around sites of JSON associative-array accesses. Reading around the code led me to the "interesting" `void[...]` "optimization" in [json.d](https://github.com/vibe-d/vibe.d/blob/master/data/vibe/data/json.d#L107).

Is this really necessary?

I put it back as `union` and the crashes seem to have disappeared.

**Comment by FeepingCreature** [comment]
https://github.com/vibe-d/vibe.d/pull/2206 PR up.

**Comment by s-ludwig** [comment]
The false pointers were caused by garbage being left in the parts of the union that were not currently used (e.g. a `bool` stored and the high-order bytes contained a pointer that was still on the stack). The `void[]` array is explicitly initialized, which causes it to be zeroed out.

I have no idea whether the union handling of the compiler has changed since then, but we'd have to be very careful when reintroducing that code. The memory leaks were quite nasty and hard to track down.

> `enum m_size = max((BigInt.sizeof+(void*).sizeof), 2);`

This is an interesting evolution that apparently nobody noticed:
```
void*[2] m_data;
void*[max((BigInt.sizeof+(void*).sizeof-1)/(void*).sizeof, 2)] m_data;
void[max((BigInt.sizeof+(void*).sizeof), 2)] m_data;
```

Should be replaced by `std.traits.Largest!(...)`.

**Comment by FeepingCreature** [comment]
Changed PR to completely zero out the union on construction.

**Comment by FeepingCreature** [comment]
I checked again and to my best ability to discern, in our usecase the union hack makes Vibe crash outright. Even in debug builds.

**Comment by Geod24** [comment]
@FeepingCreature : Still happening I suppose ?

P.S: When adding a link on the issue, press 'y' before copying the address. It gives you a link to the commit, not the branch. E.g.: https://github.com/vibe-d/vibe.d/blob/3d0558e286ef89232e277c24332d12c30de02af0/data/vibe/data/json.d#L135

**Comment by FeepingCreature** [comment]
We largely stopped using vibe in the meantime, so I don't have a repro. I have no reason to expect it to have started working.

**Comment by s-ludwig** [comment]
I'd say it makes sense to simply switch to `taggedalgebraic.taggedunion.TaggedUnion` instead of relying on a custom tagged union implementation. If the same issue happens there, it would also be a lot more valuable to have it fixed.

**Comment by s-ludwig** [comment]
Appears to be fixed by #2206
